### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by miguelhenaoa**
# Summary
- Created Shopping Cart component
- Create service to open component like modal
- Created test for shopping cart component and new methods of others components

# Trello Card
https://trello.com/c/fFS5XxSV

# Screenshot
image
image
image

### CODE REVIEW

File: src/app/components/book/book.component.html [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <div class="card">
-  <img [src]="book.image_url" alt="" class="book-cover">
+  <img id="show-book" (click)="showBook()" [src]="book.image_url" alt="" class="book-cover">
```
Comment (ebedoyaalzate) [inline src/app/components/book/book.component.html]: add alt text

File: src/app/components/book/book.component.html [hunk 2]
```diff
@@ -1,6 +1,6 @@
...
   <p class="title">{{book.title}}</p>
   <p class="author">{{book.author}}</p>
-  <ul class="btn-add"> <span>+</span> </ul>
+  <button id="btn-add" (click)="addBookToCart()" class="btn-add"> <span>+</span> </button>
```
Comment (ebedoyaalzate) [inline src/app/components/book/book.component.html]: Remember update the branches when add corrections

File: src/app/components/modal/modal.component.scss [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  position: absolute;
+  right: 0;
+  top: 0;
+  transition: opacity 250ms ease-in;
+  visibility: hidden;
+
+  > div {
+    background: white;
+    border-radius: 8px;
+    box-shadow: 0 0 8px 4px rgba(0, 0, 0, 0.1);
```
Comment (ebedoyaalzate) [inline src/app/components/modal/modal.component.scss]: Add variable

File: src/app/components/modal/modal.component.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<ng-container *ngIf="display$ | async as display">
+  <section [class.open]="display === 'open'">
+    <div>
+      <div class="dialog-header flex">
+        <span class="dialog-title">{{title}}</span>
+        <a id="btn-shopping-cart" class="btn-close" (click)="close()"></a>
```
Comment (ebedoyaalzate) [inline src/app/components/modal/modal.component.html]: Use button

File: src/app/components/modal/modal.component.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<ng-container *ngIf="display$ | async as display">
+  <section [class.open]="display === 'open'">
+    <div>
```
Comment (ebedoyaalzate) [inline src/app/components/modal/modal.component.html]: This div isn't neccesary

File: src/app/components/modal/modal.component.scss [hunk 6]
```diff
@@ -0,0 +1,59 @@
+section {
+  background: rgba(0, 0, 0, 0.2);
```
Comment (ebedoyaalzate) [inline src/app/components/modal/modal.component.scss]: Use variable

File: src/app/screens/auth/book-list/book-list.component.html [hunk 7]
```diff
@@ -1,3 +1,4 @@
+<app-header [books]="booksInCart" (showShoppingCartEvent)="showShoppingCart()"></app-header>
```
Comment (ebedoyaalzate) [inline src/app/screens/auth/book-list/book-list.component.html]: Do you really need to open the modal in the father component?

File: src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  templateUrl: './shopping-cart.component.html',
+  styleUrls: ['./shopping-cart.component.scss']
+})
+export class ShoppingCartComponent {
+
+  @Input() data: Array<Book> = [];
+
+  removeBook(book: Book) {
+    const index = this.data.findIndex(x => x === book);
+    this.data.splice(index, 1);
```
Comment (ebedoyaalzate) [inline src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts]: You could use filter in one line

**Review by ebedoyaalzate — CHANGES_REQUESTED** [review]

File: package.json [hunk 9]
```diff
@@ -11,10 +11,12 @@
   "private": true,
   "dependencies": {
     "@angular/animations": "~12.0.0",
+    "@angular/cdk": "^12.0.3",
     "@angular/common": "~12.0.0",
     "@angular/compiler": "~12.0.0",
     "@angular/core": "~12.0.0",
     "@angular/forms": "~12.0.0",
+    "@angular/material": "^12.0.3",
```
Comment (CrisVega08) [inline package.json]: Why do you need angular/material?

File: package.json [hunk 10]
```diff
@@ -11,10 +11,12 @@
   "private": true,
   "dependencies": {
     "@angular/animations": "~12.0.0",
+    "@angular/cdk": "^12.0.3",
```
Comment (CrisVega08) [inline package.json]: Why do yo need this?

File: src/app/components/header/header.component.html [hunk 11]
```diff
@@ -1,4 +1,9 @@
 <header class="header">
   <img class="logo-wolox" src="assets/images/logo_wolox.svg" alt="Logo Wolox">
-  <a id="btn-logout" class="btn btn-logout" (click)="sessionLogout()">Logout</a>
+  <div class="btns">
+    <a id="btn-shopping-cart" class="btn btn-cart" (click)="showShoppingCart()" [matBadge]="books.length" matBadgeOverlap="true" matBadgeSize="small">
```
Comment (CrisVega08) [inline src/app/components/header/header.component.html]: use button tag

File: src/app/components/header/header.component.html [hunk 12]
```diff
@@ -1,4 +1,9 @@
...
+      <img class="icon-shopping-cart" src="/assets/images/cart.svg">
+    </a>
+    <a id="btn-logout" class="btn btn-logout" (click)="sessionLogout()">Logout</a>
```
Comment (CrisVega08) [inline src/app/components/header/header.component.html]: same

File: src/app/components/header/header.component.scss [hunk 13]
```diff
@@ -1,16 +1,44 @@
 @import "src/styles/lib/colors";
 
+.btns {
+  align-content: center;
+  display: flex;
+  justify-content: space-evenly;
+}
+
+.btn-cart {
+  height: 24px;
```
Comment (CrisVega08) [inline src/app/components/header/header.component.scss]: 25px

File: src/app/components/header/header.component.scss [hunk 14]
```diff
@@ -1,16 +1,44 @@
...
+  margin-right: 30px;
+  width: 22px;
```
Comment (CrisVega08) [inline src/app/components/header/header.component.scss]: 20px

File: src/app/components/modal/modal.component.html (hunk 5 above)
Comment (CrisVega08) [reply src/app/components/modal/modal.component.html]: remove

File: src/app/components/modal/modal.component.scss [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  bottom: 0;
+  display: block;
+  left: 0;
```
Comment (CrisVega08) [inline src/app/components/modal/modal.component.scss:7]: I think you only need to properties (top or bottom and right or left)

File: src/app/components/modal/modal.component.scss [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    left: 50%;
+    padding: 30px;
+    position: absolute;
+    top: 50%;
+    transform: translate(-50%, -50%);
+    width: 900px;
+  }
+
+  &.open {
+    visibility: inherit;
```
Comment (CrisVega08) [inline src/app/components/modal/modal.component.scss:29]: is unnecessary

File: src/app/components/modal/modal.module.ts [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { ModalComponent } from './modal.component';
+
+
```
Comment (CrisVega08) [inline src/app/components/modal/modal.module.ts]: remove

**Review by CrisVega08 — CHANGES_REQUESTED** [review]

File: src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'app-shopping-cart',
+  templateUrl: './shopping-cart.component.html',
+  styleUrls: ['./shopping-cart.component.scss']
+})
+export class ShoppingCartComponent {
+
+  @Input() data: Array<Book> = [];
+
+  removeBook(book: Book) {
+    this.data.splice(this.data.findIndex(x => x === book), 1);
```
Comment (ebedoyaalzate) [inline src/app/screens/auth/components/shopping-cart/shopping-cart.component.ts]: Use filter array method for more clean code.
`this.data = this.data.filter(x => x !== book);`

**Review by ebedoyaalzate — APPROVED** [review]

**Review by CrisVega08 — APPROVED** [review]

**Pull request merged**